fixes #89 added `toObjectContaining`

hardy925 opened the pull request:
**Description**

fixes #89 adding `toObjectContaining` to revive to

```ts
{ [k: string]: SomeClass }
// or
{ [k: string]: SomeClass[] }
```

**Proposed changes in this PR**

- add `toObjectContaining`
- added tests for `toObjectContaining`
- add docs for `toObjectContaining`

**Things to look at**

- [x] Test coverage
- [x] Code Style
- [x] Documentation (`README.md`, `doc/`, `examples/`, etc..)

Commit:
fixes #89 added `toObjectContaining`

Commit:
move test coverage, adde error message for invalid date

Commit:
fix nodejs tests or new strategy

ms1111 commented on strategy/from_json/to_object_containing.ts:
```diff
@@ -0,0 +1,30 @@
+        }
+      }
+    }
+    return record;
```
What should we do if it's not null/undefined, but is not an object here? (A string or a number.) Should we throw an error or let it through?

hardy925 replied:
added errors!, thanks!

ms1111 reviewed (dismissed):
Nice!

Commit:
issue 89 prr

Commit:
issue 89 prr

Commit:
throw errors  for invalid values

Commit:
remove console log

Commit:
update toJSON to run on empty arrays properly

Commit:
update toJSON to run on empty arrays properly

Commit:
update toJSON to run on empty arrays properly

Commit:
update toJSON to run on empty arrays properly

Commit:
update toJSON to run on empty arrays properly

Commit:
update error message name

Commit:
toObjectContaining throws is array sub-value values are not [object Object]

ChrisDufourMB commented on strategy/from_json/to_object_containing.ts:
```diff
@@ -0,0 +1,62 @@
+
+/** revive data from `{k: v}` using `fromJSON` on a subclass type `v` */
+export function toObjectContaining<T>(
+  type: T & { new (): Serializable },
```
I am a bit concerned with the `{ new (): Serializable }` type here as it forces the Object class to not implement a constructor with arguments, which may not be the case for some classes.

![image](https://user-images.githubusercontent.com/49878815/101657211-a01c5b00-3a11-11eb-9980-e457e9a6b920.png)
^ `toObjectContaining` won't accept `SomeClass` as an argument becase it doesn't have a no args constructor

I think an initializer function could work around this:
```ts
toObjectContaining(() => new SomeClass("defaultValue"))
```

hardy925 replied:
`toSerializable` would also have the same issue

ChrisDufourMB replied:
Hmm, well if it's at least consistent we can probably just leave it for now.

We may want to look into an overload or something at some point though so we can support constructors with args. This would help when implementors want to gradually move thier classes over to ts_serialize, rather than having to rewrite all their classes without no-args constructors.

hardy925 replied:
fixed it for both `toObjectContaining` an `toSerializable`

ChrisDufourMB commented on strategy/from_json/to_object_containing.ts:
```diff
@@ -0,0 +1,62 @@
+      throw Error(ERROR_TO_OBJECT_CONTAINING_USE_TO_SERIALIZABLE);
+    }
+
+    if (Object.prototype.toString.call(value) !== "[object Object]") {
```
Is this the best way to check is a given value is a POJO object? Maybe a comment here could help.

hardy925 replied:
Yea it is, I know it's expensive

ChrisDufourMB commented on strategy/from_json/to_object_containing.ts:
```diff
@@ -0,0 +1,62 @@
+    const record: Record<string, Serializable | Serializable[] | null> = {};
+    // check that JSONValue is something we can deal with
+    // but mostly to make the type checker happy
+    if (typeof value === "object" && !Array.isArray(value)) {
```
```suggestion
    if (typeof value === "object") {
```
The earlier check should have caught any arrays

hardy925 replied:
oh ya, added the above check pretty late, thanks

ChrisDufourMB commented on strategy/from_json/to_object_containing.ts:
```diff
@@ -0,0 +1,62 @@
+            if (
+              Object.prototype.toString.call(v) !== "[object Object]"
+            ) {
+              throw Error(ERROR_TO_OBJECT_CONTAINING_INVALID_SUB_VALUE);
```
```suggestion
              throw new Error(ERROR_TO_OBJECT_CONTAINING_INVALID_SUB_VALUE);
```

Looks like there is a missing `new` keyword?

ChrisDufourMB commented on strategy/from_json/to_object_containing.ts:
```diff
@@ -0,0 +1,62 @@
+          continue;
+        }
+
+        if (Object.prototype.toString.call(value[prop]) !== "[object Object]") {
```
This pattern is used in a few places, maybe a util function would be useful?

ChrisDufourMB reviewed (dismissed):
Blocking because `serializable.ts` changes break handling empty arrays

The `toObjectContaining` work looks good but enforces a specific constructor pattern which may make implementing it difficult in some circumstances

hardy925 commented:
> Blocking because serializable.ts changes break handling empty arrays

Am I missing a test?

ChrisDufourMB commented on serializable.ts:
```diff
@@ -193,7 +193,7 @@ function fromJSON<T>(
         } = serializablePropertyMap.getBySerializedKey(key) || {};
 
-        const processedValue: unknown = Array.isArray(value)
+        const processedValue: unknown = Array.isArray(value) && value.length
```
@shardyMBAI Sorry, seems that this review comment didn't actually get added somehow

Adding in this constraint causes empty arrays to not be serialized into arrays:
```
test({
  name: "Revives an empty array of `type`",
  fn() {
    class OtherClass extends Serializable {
      @SerializeProperty()
      id!: number;
    }
    class Test extends Serializable {
      @SerializeProperty({
        fromJSONStrategy: (v) => new OtherClass().fromJSON(v),
      })
      array!: OtherClass[];
    }
    const testObj = new Test().fromJSON(
      `{"array":[]}`,
    );

    assert(Array.isArray(testObj.array)); // Test fails here - Test { array: OtherClass {} }
    assertEquals(testObj.array.length, 0);
  },
});
```

hardy925 replied:
damnit, thanks!

WillWang-MindBridge commented on strategy/from_json/date.ts:
```diff
@@ -1,12 +1,13 @@
   return function _createDateStrategy(input: JSONValue): Date {
     if (typeof input !== "string" || !regex.exec(input)) {
-      throw new Error("invalid date");
+      throw new Error(ERROR_INVALID_DATE);
```
I see we are only using the passed in regex for validation. Should you extract the 3 numbers and use that to build the date instead?

hardy925 replied:
Hey, 
There are probably some improvements that could be made here, but the goal is to parse a string and convert that to a Date if it passes. We're not sure what string or regex we'll get as well. Is there a bug it would be fixing?

WillWang-MindBridge replied:
regex: `^Im going to shoot my foot$`
input: `Im going to shoot my foot`
it's gonna fail at `new Date(input)`
I agree it's not a main concern

hardy925 replied:
We can't fix issues in projects that use this library, and if they are inputting a "string like this" expecting a Date, we can reply on native JS to let them know they messed up.

hardy925 replied:
I have this test passing

```ts

test({
  name: "createDateStrategy - throws on invalid dates",
  fn() {
    class Test extends Serializable {
      @SerializeProperty({
        fromJSONStrategy: createDateStrategy(/^Im going to shoot my foot$/),
      })
      date!: Date;
    }
    try {
      new Test().fromJSON(`{"date":"Im going to shoot my foot"}`);
      fail("Non date string did not error");
    } catch (error) {
      assertEquals(error.message, ERROR_INVALID_DATE);
    }
  },
});
```

Commit:
prr

Commit:
allow constructor arguments for toObjectContaining an toSerializalbe

Commit:
allow constructor arguments for toObjectContaining an toSerializalbe -- missing file

Commit:
better typing

Commit:
update createDateStrategy to throw on an valid date

Commit:
add fromObjectContaining strategy

Commit:
changelog

ms1111 commented on strategy/from_json/date_test.ts:
```diff
@@ -20,7 +21,25 @@ test({
 
 test({
-  name: "Will not serialize non date strings",
+  name: "createDateStrategy - Will not serialize non date strings",
```
```suggestion
  name: "createDateStrategy - Will not deserialize non date strings",
```

ms1111 commented on strategy/from_json/date_test.ts:
```diff
@@ -20,7 +21,25 @@ test({
+});
+
+test({
+  name: "iso8601Date - Will not serialize non date strings",
```
```suggestion
  name: "iso8601Date - Will not deserialize non date strings",
```

ms1111 commented on strategy/_utils.ts:
```diff
@@ -0,0 +1,26 @@
+function isNewable<T>(
+  type: NewSerializable<T> | FunctionSerializable,
+): type is NewSerializable<T> {
+  return !!type.name;
```
Neat.

I did some Googling, what would you think about https://stackoverflow.com/a/57862312/810525
```
function isConstructor(func) {
    return (func && typeof func === "function" && func.prototype && func.prototype.constructor) === func;
}
```
The relevant bit of that is probably `func.prototype && (func.prototype.constructor === func)`.

ms1111 commented on strategy/from_json/to_object_containing.ts:
```diff
@@ -0,0 +1,55 @@
+        // null is a JSONValue
+        if (value[prop] === null) {
+          record[prop] = null;
+          continue;
+        }
+        if (Array.isArray(value[prop])) {
+          record[prop] = (value[prop] as JSONArray).map((v: JSONValue) => {
+            if (!isObject(v)) {
+              throw new Error(ERROR_TO_OBJECT_CONTAINING_INVALID_SUB_VALUE);
+            }
+            return getNew(type).fromJSON(v);
+          });
+          continue;
```
How would you feel about changing the `if / continue / if / continue` to `if / else if ... / else if ... / else`? I watched a talk (from a C++ conference of all things) where someone was explaining why that can be easier to read (makes the control structure obvious just by looking at the indentation, basically).

hardy925 replied:
I am a fan of exit early exit often. We're not doing crazy processing so the shortcircuit seems appropriate.

ms1111 commented on strategy/to_json/from_oject_containing_test.ts:
```diff
@@ -0,0 +1,68 @@
+// Copyright 2018-2020 Gamebridge.ai authors. All rights reserved. MIT license.
```
Typo in filename - from_oject -> from_object

ms1111 commented on strategy/to_json/from_object_containing.ts:
```diff
@@ -0,0 +1,16 @@
+): JSONObject {
+  const record: Record<string, JSONValue> = {};
+  for (const prop in value) {
+    record[prop] = Array.isArray(value[prop])
+      ? (value[prop] as Serializable[]).map((v) => v.tsSerialize())
+      : (value[prop] as Serializable).tsSerialize();
```
I thought this already happened by default. Maybe just for arrays? I figure if we have the objects in memory, we know if they are Serializable, so the default strategy should be to serialize _really hard_. Recursively go through each value in an embedded array or object and `tsSerialize()` it.

But again I thought we already did that by default (at least one level deep). Maybe we only did it for arrays?

hardy925 replied:
we do this by default with toJSON but that returns a string and parsing causes escaping, by using `tsSerialize` we are avoiding that issue.

ms1111 reviewed (commented):
I wonder if fromObjectContaining() needs to exist or if we should just always do it. Looks really good other than that.

hardy925 commented:
Added requested changes, also added `fromObjectContaining` as a `toJSONstrategy`,  @ChrisDufourMB your change also has type protection.

<img width="972" alt="Screen Shot 2020-12-10 at 11 59 22 PM" src="https://user-images.githubusercontent.com/27963753/101865242-c04c3700-3b43-11eb-9048-1e0733f8c883.png">

Commit:
Update strategy/from_json/date_test.ts

Co-authored-by: Michael Smith <msmith@mindbridge.ai>

Commit:
Update strategy/from_json/date_test.ts

Co-authored-by: Michael Smith <msmith@mindbridge.ai>

Commit:
file name change

Commit:
Merge branch 'iss-89' of github.com:GameBridgeAI/ts_serialize into iss-89

Commit:
isNewable change

Commit:
isNewable change

Commit:
use `continue` only to shortcircut

Commit:
revert else ifs

Commit:
revert else ifs

Commit:
nodejs expots

Commit:
add interfaces for all public methods on serializable

Commit:
add interfaces for all public methods on serializable

Commit:
another test;

Commit:
remove `only` from test

Commit:
add  du4s test, deno@1.6.0 sdt@0.80.0;

ChrisDufourMB commented on strategy/_utils.ts:
```diff
@@ -0,0 +1,27 @@
+}
+
+/** get new strategy type arguments */
+export function getNew<T>(
```
💯 , may look into adding this to the polymorphic initializer too
